lib/posix/semaphore.c

```diff
@@ -53,17 +53,20 @@ static void nsem_cleanup(struct nsem_obj *nsem)
 	if (nsem != NULL) {
 		if (nsem->name != NULL) {
 			k_free(nsem->name);
-			nsem->name = NULL;
 		}
 		k_free(nsem);
-		nsem = NULL;
 	}
 }
 
 /* Remove a named semaphore if it isn't unsed */
-static void nsem_remove_if_unused(struct nsem_obj *nsem)
+static void nsem_unref(struct nsem_obj *nsem)
 {
+	nsem->ref_count -= 1;
+	__ASSERT(nsem->ref_count >= 0, "ref_count may not be negative");
+
 	if (nsem->ref_count == 0) {
+		__ASSERT(nsem->name == NULL, "ref_count is 0 but sem is not unlinked");
```

Comment:
@npitre This means that the following behavior is guaranteed only when `CONFIG_ASSERT` is enabled, is that intended?

> If the semaphore has not been removed with a successful call to [sem_unlink()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html), then [sem_close()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_close.html) has no effect on the state of the semaphore.

i.e.: a named semaphore can be closed without being unlinked:

```
operation                refcount change          refcount state
----------------------------------------------------------------

sem_open()               = 2 (created)            2
sem_close()              - 1                      1
sem_close()              - 1                      0
```